src/lib/core/Optional.h

```diff
@@ -37,77 +37,119 @@ struct NullOptionalType
 };
 constexpr NullOptionalType NullOptional{};
 
+namespace detail {
+
+/// Base class that handles destruction for Optional<T>
+template <class T, typename Enable = void>
+struct OptionalBase
+{
+    bool mHasValue;
+    union
+    {
+        char mEmpty;
+        T mValue;
+    };
+
+    constexpr OptionalBase() : mHasValue(false), mEmpty(0) {}
+
+    ~OptionalBase()
+    {
+        // NOLINTNEXTLINE(clang-analyzer-core.uninitialized.Branch): mValue is active when mHasValue
+        if (mHasValue)
+        {
+            mValue.~T();
+        }
+    }
+};
+
+template <class T>
+struct OptionalBase<T, std::enable_if_t<std::is_trivially_destructible<T>::value>>
+{
+    bool mHasValue;
+    union
+    {
+        char mEmpty;
+        T mValue;
+    };
+
+    constexpr OptionalBase() : mHasValue(false), mEmpty(0) {}
+
+    ~OptionalBase() = default;
+};
+
+} // namespace detail
+
 /**
  * Pairs an object with a boolean value to determine if the object value
  * is actually valid or not.
  */
 template <class T>
-class Optional
+class Optional : protected detail::OptionalBase<T>
```

Comment:
Could we directly wrap a `std::optional` instead now that we are C++17 and also mark this deprecated for std::optional directly?

Reply:
We could, but it would be a more substantial change. Maybe the migration to std::optional is something we should tackle separately?

Reply:
Turns out with correct optimization settings the destructors of chip::Optional get optimized away fine as is even if they're not technically trivial, so I'm leaving it as is.